Add the frozen tier node role and ILM phase

dakrone opened the pull request:
This commit adds the `data_frozen` node role as part of the formalization of data tiers. It also
adds the `"frozen"` phase to ILM, currently allowing the same actions as the existing cold phase.

The frozen phase is intended to be used for data even less frequently searched than the cold phase,
and will eventually be loosely tied to data using partial searchable snapshots (as oppposed to full
searchable snapshots in the cold phase).

Relates to #60848

dakrone committed:
Add the frozen tier node role and ILM phase

This commit adds the `data_frozen` node role as part of the formalization of data tiers. It also
adds the `"frozen"` phase to ILM, currently allowing the same actions as the existing cold phase.

The frozen phase is intended to be used for data even less frequently searched than the cold phase,
and will eventually be loosely tied to data using partial searchable snapshots (as oppposed to full
searchable snapshots in the cold phase).

Relates to #60848

elasticmachine commented:
Pinging @elastic/es-core-features (Team:Core/Features)

dakrone commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshots.java:
```diff
@@ -182,9 +182,11 @@
      */
-    public static final String DATA_TIERS_PREFERENCE = String.join(",", DataTier.DATA_COLD, DataTier.DATA_WARM, DataTier.DATA_HOT);
+    public static final String DATA_TIERS_PREFERENCE =
+        String.join(",", DataTier.DATA_COLD, DataTier.DATA_FROZEN, DataTier.DATA_WARM, DataTier.DATA_HOT);
```
This may seem backward, but it was at @DaveCTurner's recommendation

jasontedor replied:
I think it's worth a comment in the code as to why. We certainly don't want this inadvertently changed, or someone seeing it in the code, having to a `git blame`, track down this PR, and find this GitHub comment. 😇

jasontedor replied:
I think we need to separate this preference from the preference used in the mount action.

For the mount action, I think we want two preferences:

The preference when the `storage` setting is `shared_cache` which has the preference frozen -> cold -> warm -> hot
The preference when the `storage` setting is `full_copy` which has the preference cold -> warm -> hot

And for the cache, we'd use the preference as you have here.

jasontedor commented on docs/reference/ilm/ilm-index-lifecycle.asciidoc:
```diff
@@ -6,13 +6,16 @@
 * **Cold**: The index is no longer being updated and is seldom queried. The
 information still needs to be searchable, but it's okay if those queries are
 slower.
+* **Frozen**: The index is no longer being updated and is even more seldom queried. The
```
Hmm...`more seldom` is an odd phrasing, in the way that a double negative is. How about rewriting `seldom queried` in `Cold` to `queried infrequently` and in `Frozen` to `queried rarely`?

jasontedor commented on docs/reference/datatiers.asciidoc:
```diff
@@ -74,12 +75,23 @@ For resiliency, indices in the warm tier should be configured to use one or more
 
 Once data is no longer being updated, it can move from the warm tier to the cold tier where it
-stays for the rest of its life.
+stays while being seldomly queried.
```
How about `queried infrequently`?

dakrone committed:
Spotless ಠ_ಠ

jasontedor commented on docs/reference/datatiers.asciidoc:
```diff
@@ -74,12 +75,23 @@ For resiliency, indices in the warm tier should be configured to use one or more
+[[frozen-tier]]
+=== Frozen tier
+
+Once data is no longer being queried, or being quered very seldomly, it may move from the cold tier
```
`quered` -> `queried` (typo)

And maybe `queried rarely`?

jasontedor commented on docs/reference/datatiers.asciidoc:
```diff
@@ -11,6 +11,7 @@ and hold your most recent, most-frequently-accessed data.
 * <<warm-tier, Warm tier>> nodes hold time series data that is accessed less-frequently
 and rarely needs to be updated.
 * <<cold-tier, Cold tier>> nodes hold time series data that is accessed occasionally and not normally updated.
```
`accessed infrequently`?

jasontedor commented on docs/reference/datatiers.asciidoc:
```diff
@@ -11,6 +11,7 @@ and hold your most recent, most-frequently-accessed data.
 * <<warm-tier, Warm tier>> nodes hold time series data that is accessed less-frequently
 and rarely needs to be updated.
 * <<cold-tier, Cold tier>> nodes hold time series data that is accessed occasionally and not normally updated.
+* <<frozen-tier, Frozen tier>> nodes hold time series data that is accessed very seldom and never updated.
```
`accessed rarely`?

jasontedor commented on docs/reference/glossary.asciidoc:
```diff
@@ -207,6 +207,19 @@ Frozen indices use a memory-efficient shard implementation and throttle searches
+[[glossary-frozen-phase]] frozen phase ::
+// tag::frozen-phase-def[]
+The fourth possible phase in the <<glossary-index-lifecycle,index lifecycle>>.
+In the frozen phase, an index is no longer updated and very seldom queried.
```
`queried rarely`?

jasontedor commented on docs/reference/glossary.asciidoc:
```diff
@@ -207,6 +207,19 @@ Frozen indices use a memory-efficient shard implementation and throttle searches
+[[glossary-frozen-tier]] frozen tier::
+// tag::frozen-tier-def[]
+A <<glossary-data-tier, data tier>> that contains nodes that hold time series data
+that is accessed occasionally and not normally updated.
```
I don't think we want `occasionally` here, maybe `rarely`?

jasontedor commented on server/src/main/java/org/elasticsearch/indices/recovery/RecoverySettings.java:
```diff
@@ -43,9 +43,8 @@
-                if ((dataRoles.size() > 1 || dataRoles.get(0).roleName().equals("data_cold") == false) ||
-                    roles.contains(DiscoveryNodeRole.MASTER_ROLE)) {
-                    // if the node is not a dedicated cold node, use the default
+                if (dataRoles.stream().allMatch(dn -> dn.roleName().equals("data_cold") || dn.roleName().equals("data_frozen")) == false) {
```
Good catch.

jasontedor reviewed (commented):
LGTM. The only question is around the descriptions of the querying levels in the tiers. I offered a suggestion of `cold -> infrequently, frozen -> rarely`, but I am open to other suggestions. The `seldom`, `more seldom` reads oddly to me.

dakrone committed:
Address documentation review

dakrone committed:
Fix broken doc link (typo)

jasontedor commented on docs/reference/glossary.asciidoc:
```diff
@@ -207,6 +207,19 @@ Frozen indices use a memory-efficient shard implementation and throttle searches
+[[glossary-frozen-phase]] frozen phase ::
+// tag::frozen-phase-def[]
+The fourth possible phase in the <<glossary-index-lifecycle,index lifecycle>>.
+In the frozen phase, an index is no longer updated and very queried rarely.
```
`very queried rarely` -> `queried rarely`

jasontedor commented on docs/reference/datatiers.asciidoc:
```diff
@@ -10,7 +10,8 @@ typically share the same hardware profile:
 and rarely needs to be updated.
-* <<cold-tier, Cold tier>> nodes hold time series data that is accessed occasionally and not normally updated.
+* <<cold-tier, Cold tier>> nodes hold time series data that is accessed rarely and not normally updated.
+* <<frozen-tier, Frozen tier>> nodes hold time series data that is accessed very seldom and never updated.
```
`accessed very seldom` -> `accessed rarely`

jasontedor commented on docs/reference/datatiers.asciidoc:
```diff
@@ -10,7 +10,8 @@ typically share the same hardware profile:
 * <<warm-tier, Warm tier>> nodes hold time series data that is accessed less-frequently
 and rarely needs to be updated.
-* <<cold-tier, Cold tier>> nodes hold time series data that is accessed occasionally and not normally updated.
+* <<cold-tier, Cold tier>> nodes hold time series data that is accessed rarely and not normally updated.
```
`accessed rarely` -> `accessed infrequently`

jasontedor reviewed (commented):
A few nits 😇.

dakrone committed:
Pick more better words

dakrone committed:
Fix cat nodes tests by adding missing abbreviation

dakrone committed:
More cat nodes abbreviation fixes

jasontedor reviewed (dismissed):
I left one more comment about a code comment, but I don't think we need another round.

dakrone committed:
Make tier preference depend on mount type

A full copy mount prefers cold -> warm -> hot
A partial mount prefers frozen -> cold -> warm -> hot

dakrone committed:
Fix autoscaling yml test

jasontedor reviewed (approved):
LGTM.

dakrone committed:
Fix test

jasontedor commented:
I think we missed one place in the docs which is docs/reference/modules/node.asciidoc. Leaving this here as a reminder for one of us to pick this up on Monday.

jasontedor commented:
The above [comment](https://github.com/elastic/elasticsearch/pull/68605#issuecomment-774491094) will be addressed by #68713.
